presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java

```diff
@@ -0,0 +1,261 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.QueryId;
+import com.facebook.presto.spi.predicate.Domain;
+import com.facebook.presto.spi.predicate.TupleDomain;
+import com.facebook.presto.sql.planner.DomainTranslator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.relational.RowExpression;
+import com.facebook.presto.sql.tree.DynamicFilterExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.LogicalBinaryExpression;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.facebook.presto.util.RowExpressionConverter;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.http.client.FullJsonResponseHandler;
+import io.airlift.log.Logger;
+
+import javax.ws.rs.core.Response;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.sql.relational.Expressions.call;
+import static com.facebook.presto.sql.relational.Signatures.logicalExpressionSignature;
+import static java.util.Objects.requireNonNull;
+
+public class DynamicFilterCollector
```

Comment:
`DynamicFilterCollector` seems not to be closed when `ScanFilterAndProject` operator closes. Can it query master after the `SFP` finishes?

Reply:
yeah currently it can ... it does close itself after retrying all attempts but it doesn't know if SFP finished. basically there is no close function in PageFilter that can be propagated to collector.

Reply:
Maybe it could be closed on `SFP::close()`. Alternatively, we could add `close` method to `PageFilter`.